isis/src/base/apps/csminit/csminit.cpp

```diff
@@ -0,0 +1,305 @@
+          QStringList modelSpec = {pluginName, modelName, QString::fromStdString(nitf21Isd.format())};
+          possibleModels.append(modelSpec);
+          continue; // If the NITF 2.1 ISD works, don't check the others
+        }
+      }
+    }
+
+    if (possibleModels.size() > 1) {
+      QString message = "Multiple models can be created from the ISD [" + isdFilePath + "]. "
+                        "Re-run with the PLUGINNAME and MODELNAME parameters. "
+                        "Possible plugin & model names:\n";
+      for (const QStringList &modelSpec : possibleModels) {
+        message += "Plugin [" + modelSpec[0] + "], Model [" + modelSpec[1] + "]\n";
+      }
+      throw IException(IException::User, message, _FILEINFO_);
+    }
+
+    if (possibleModels.empty()) {
+      QString message = "No loaded model could be created from the ISD [" + isdFilePath + "]."
+                        "Loaded plugin & model names:\n";
+      for (const csm::Plugin * plugin : csm::Plugin::getList()) {
+        QString pluginName = QString::fromStdString(plugin->getPluginName());
+        for (size_t modelIndex = 0; modelIndex < plugin->getNumModels(); modelIndex++) {
+          QString modelName = QString::fromStdString(plugin->getModelName(modelIndex));
+          message += "Plugin [" + pluginName + "], Model [" + modelName + "]\n";
+        }
+      }
+      throw IException(IException::User, message, _FILEINFO_);
+    }
+
+    // If we are here, then we have exactly 1 model
+    QStringList modelSpec = possibleModels.front();
+    if (modelSpec.size() != 3) {
+      QString message = "Model specification [" + modelSpec.join(" ") + "] has [" + modelSpec.size() + "] elements "
+                        "when it should have 3 elements.";
+      throw IException(IException::Programmer, message, _FILEINFO_);
+    }
+    const csm::Plugin *plugin = csm::Plugin::findPlugin(modelSpec[0].toStdString());
+    csm::Model *model;
+    csm::Isd fileIsd(isdFilePath.toStdString());
+    csm::Nitf21Isd nitf21Isd(isdFilePath.toStdString());
+    if (modelSpec[2] == QString::fromStdString(fileIsd.format())) {
+      model = plugin->constructModelFromISD(fileIsd, modelSpec[1].toStdString());
+    }
+    else if (modelSpec[2] == QString::fromStdString(nitf21Isd.format())) {
+      model = plugin->constructModelFromISD(nitf21Isd, modelSpec[1].toStdString());
+    }
+    else {
+      QString message = "Invalid ISD format specifications [" + modelSpec[2] + "].";
+      throw IException(IException::Programmer, message, _FILEINFO_);
+    }
+
+    string modelState = model->getModelState();
+
+    // Add the TargetName to the instrument group, if specified:
+    if (ui.WasEntered("TARGETNAME")) {
+      if (!cube->hasGroup("Instrument")) {
+        cube->putGroup(PvlGroup("Instrument"));
+      }
+      PvlGroup &instrumentGroup = cube->group("Instrument");
+      if (instrumentGroup.hasKeyword("TargetName")) {
```

Comment:
Instead of deleting and re-adding the targetname keyword, couldn't we overwrite the value of the keyword in one line?

Reply:
addKeyword with Pvl::Replace would be better here.